src/AWS/Lambda/Context.hs

```diff
@@ -83,8 +82,22 @@ class HasLambdaContext r where
 instance HasLambdaContext LambdaContext where
   withContext = const
 
-instance DefConfig LambdaContext where
-  defConfig = LambdaContext "" "" 0 "" "" "" "" "" (posixSecondsToUTCTime 0) Nothing Nothing
+-- TODO: This sticks around for both backwards compatibility and the lack of a
```

Comment:
You could use the `Default` class, but I generally don't like `data-default`. It might be appropriate here because you've previously said how much you want to avoid backwards compat changes?

I think a lot of trouble goes away if you avoid `ReaderT` and `MonadReader` as your way of passing the `LambdaContext` on each invocation, and instead ask the caller to provide a function `LambdaContext -> m whatever`. If the caller doesn't want to use the context, he or she can use `const` to hide it, and the "is it here or not?"-ness of `mRuntimeWithContext` etc goes away if you was for `MonadWhatever m => (LambdaContext -> event -> m Result) -> m ()`

Reply:
Hey thanks for weighing in on this, it's an area where I wasn't feeling that confident in, so I appreciate the review.

The `DefConfig` served two purposes 1) it was a requirement for envy's FromEnv, which was used to retrieve the `StaticContext`.  2) When using `runReaderT`, you needed a value for the environment, and this worked.  On reflection (1) just wasn't doing all that much and (2) was just maybe never really a good solve for an issue that I still don't know how to get to the core of.

And yeah, not only am I trying to not issue breaking changes, but this one theoretically _is_ because it drops an instance.  If someone uses `defConfig` exported anywhere else but this module, they'll be broken.  So in the very strict sense, this already _should_ be considered a breaking change.

So, I'm not sure that we'd want to use the `Default` class, because as a replacement, it still brings along some of these issues that I'd rather fully solve.

Your point on the `LambdaContext` is interesting.  My initial thought was that, "hey, `LambdaContext` can just go right in the monad with ReaderT!" and that it would super be convenient.  In reality, I'm not sure that's the case.  It might be my own shortcomings, but the `ReaderMonad` can still be quite a pain to combine environments.  I typically don't use the `LambdaContext` and find myself tediously declaring useless no-ops for `HasLambdaContext` just to get things to compile.  Other projects that don't play nice with sharing their environment with `Has` like classes get especially tricky.

Also, if we've just fundamentally gone about the interface wrong, this _is_ where I tend to lean towards a breaking change.  I'd still prefer to add new handlers and preserve old ones if they can reasonably live along side each other, but yeah, if there's a better way, then there's a better way.

I have thought to myself that maybe there's a place for a runtime that just doesn't have `LambdaContext`.  If we _knew_ you didn't need it, then we don't have to do any work at all to parse it.  I think the combinator strategy would help making that maintainable.

Thinking more, I see that `DefConfig` is unchanged in all versions of envy.  It may make sense to just continue supporting it.  `envy` remains a dependency, but the key goal of broader compatibility is addressed.

Pulling the `LambdaContext` back out as a parameter does seem like it would solve some things, however, it seems like something that _should_ be able to live in the monad itself.  Maybe the issue is trying to use the `MonadReader`?  This might be crazy, but what if we had a `LambdaContextT` and `MonadLambdaContext` constraint?  Then they wouldn't "bump into" other `Reader` contents.  The "O(n^2)" instances is a bit annoying, but we could probably support all the primary ones.

Curious to get your take on all this?

Reply:
Thoughts, in no particular order:

* I'd love to see `envy` removed from `build-depends` all together. We have had version conflicts between `hal`'s constraints and something else in our code making an `envy` constraint unsatisfiable. We worked around this by removing `envy` but I think its continued presence in `hal` hurts more than it helps.
* The fundamental problem with trying to smuggle the `LambdaContext` into any sort of Monad `m` means you end up with a function like (ignoring constraints) `(input -> m output) -> m ()`, and this is just fundamentally telling fibs in the type signature. Specifically, there's no sensible `LambdaContext` that can go into `m` when you're not actually handling a request.
* A custom transformer doesn't solve this problem. You _could_ write something with type `(input -> LambdaContextT m output) -> m ()` but then you're forced into concretising part of the transformer stack, and you have to write instances to lift everything through `LambdaContextT`. I **strongly** recommend against doing this: your callers would lose the ability to interoperate with things that use `MonadReader`, and there's no good places to put lifting instances for anyone else's nonstandard transformers. Either you have to introduce spurious dependencies or ask them to depend on `hal`. (Concrete example of this sort of problem: the `mmorph`-flavoured swizzling to get `amazonka`'s `AWST` and `katip`'s `KatipContextT` to play together is a bit annoying and subtle.
* I have really enjoyed working with some of the simpler runtime functions, because they allow the caller to stack on transformers where he or she wants to. Consider `ioRuntime :: (FromJSON event, ToJSON result) => (event -> IO (Either String result)) -> IO ()`. This can be easily used directly, or the caller can stack an `ExceptT` like: `ioRuntime $ \event -> runExceptT $ do ...`.
* Passing the `LambdaContext` as a parameter makes it consistent with the other `...WithContext` functions, which seems like a predictability win for your callers.
* Remark: because it's _relatively_ easy to stack transformers, `readerTRuntime` is actually redunant. A caller can put the context, event, or both into a `ReaderT` by hand if so inclined. I'd consider documenting these idioms and deprecating `readerTRuntime` but keep it around indefinitely. (It's something like: `flip $ \input -> runReaderT $ do ...`, so a little gross compared to the `ExceptT` example, but doable.

Overall, I think the strength of `hal` is its simplicity (I am stoked at the amount of code I'm deleting as I migrate off `aws-lambda-haskell-runtime`), and I think passing the `LambdaContext` as a function parameter is the best way to handle it.

Reply:
P.S.: Thanks for the invitation to collaborate. I've enjoyed our discussions.

Reply:
This is great stuff!

I definitely was aiming for full envy removal--and it's still on the table.  My expectation is, that only using `DefConfig` solves the constraint headaches.  hal's only constraint after this would be >= 1.0 (TODO) as it now works with all 1.x and 2.x.  The true dependency is so minimal, that this should continue to be the case for some time.  The lingering issue is that envy has an incorrect lower bound.  They accepted a PR, but still need to publish a revision.  If this takes too long then it's definitely gone all together.

In general, it's going to be gone on the next breaking change.  Before breaking, I think it's at least having the dust settle on the `Context` direction, as it may be a two birds with one stone scenario.

The ease of transformer stacking is also pretty interesting.  In this case it's like picking the closest runtime to what you want, and then stacking a little to get exactly what you want, yeah?  It may be a pattern worth explicitly documenting to guide people towards further refinement of runtimes.

> The fundamental problem with trying to smuggle the LambdaContext into any sort of Monad m means you end up with a function like (ignoring constraints) (input -> m output) -> m (), and this is just fundamentally telling fibs in the type signature. Specifically, there's no sensible LambdaContext that can go into m when you're not actually handling a request.

Nailed it, this is exactly it.

> You could write something with type `(input -> LambdaContextT m output) -> m ()`

`(input -> LambdaContextT m output) -> m ()` is essentially what I was thinking (but hadn't quite made the leap!).  In this case, the nice thing is that the runtime client can add this to the stack, and then the fib of context availability is eliminated.

> there's no good places to put lifting instances for anyone else's nonstandard transformers

The instances is definitely an issue.  If a user is defining they're on transformers, then mostly the issue is boilerplate (that they've already opted into), but if a user has to get two custom transformers to play nicely together, then yeah, that's a lot bigger of an issue.  Even if they understand what is an advanced issue, you end up with orphaned instances, right?

> the mmorph-flavoured swizzling to get amazonka's AWST and katip's KatipContextT to play together is a bit annoying and subtle

Interestingly enough though, partly the idea of moving away from `Reader` is the issues I faced with amazonka's `AWST`.  Since it doesn't have `Has` like instances, it's not very good at sharing the `Reader`'s type.  Maybe `magnify` solves this better than I realized?  Probably need some docs on how to handle these situations in either way.  Isn't the `KatipContextT` actually a similar issue, where the problem is that (by default) it's an alias of `ReaderT` so it doesn't share nicely with `AWST`?

Having been playing with polysemy/Effects, I'm wishing for something like that with type application and multiple `Reader` constraints.  It just may be that I'm wanting more than mtl can provide.

That's great to hear on it living up to its goal of simplicity!  This part has definitely been a sore thumb in that category, so I'm eager to solve it--and having it be a parameter be it.

Reply:
> The ease of transformer stacking is also pretty interesting. In this case it's like picking the closest runtime to what you want, and then stacking a little to get exactly what you want, yeah? It may be a pattern worth explicitly documenting to guide people towards further refinement of runtimes.

Yes, it lets the caller choose to use `mtl` or not. It would even be possible to plug in something like `freer-simple` in. This is another benefit of providing simple interfaces (e.g.: function parameter instead of `ReaderT`, returning `IO (Either ...)` instead of `ExceptT e IO ()`). Agree that this is worth documenting.

Re: `KatipContextT`/`AWST`: from memory, `KatipContextT` is implemented in terms of `ReaderT` but hides this fact. It provides a lifting instance instead. `AWST` provides a `MonadReader` instance that exposes its `Env`. We wound up creating our own `newtype` `Monad` that provides `MonadAWS` and `MonadKatipContext` so we could stack transformers on top of it.

I still think that the best way to avoid these issues is to be as agnostic as possible about the monad the caller uses, and that `LambdaContext`-as-parameter fits this philosophy and the philosophy of the library as a whole. If, however, you do end up introducing a transformer, try to give it instances from `mmorph` if at all possible. That was the only way I was able to stack and write both `MonadAWS` and `MonadKatipContext` instances for the newtyped monad.

P.S.: The least bad solution to MTL's "multiple `Reader` effect" problem that I know of is classy lenses to share the context, but there's a few competing libraries for capturing the "has" pattern and no clear winner AFAICT.

Reply:
Okay, yeah, I played around a bit with adding a `LambdaContextT` transformer and now fully see what you're saying.  There's just no practical way for this package to make the transformer instances of all possible Monads of interest.  And if you run into a missing instance, you basically need to start from scratch (like what you described).

It does maybe seem worth it to add in a `mRuntime??? :: (LambdaContext -> a -> m b) -> m ()` because that can then be the parent of all other runtimes, as it eliminates the `MonadReader` constraint all together, and never pretends to have the `LambdaContext` when it doesn't.

That could be added today with its own name, and then perhaps `mRuntimeWithContext` and `readerTRuntime` as they are today could be deprecated.  For those who still want it in the Monad, but don't have competitors for the Reader's environment, then perhaps adding `readerTRuntime??? :: a -> ReaderT LambdaContext m b -> m ()` makes sense?

Ultimately, these approaches get away from the lie today that the `LambdaContext` is always available when its not.

And yeah, the classy lens approach is more-or-less the inspiration behind the `HasLambdaContext`, and it does solve integrating with things like amazonka.  It's boilerplate--but probably less than most options.

Reply:
For consistency with all the other `...WithContext` names in this library, I think that the "best" name for the function with type `(Constraints) => (LambdaContext -> event -> m result) -> m ()` is `mRuntimeWithContext`. There's the small problem that it's already taken.

What to do about this? I see a few options:

1. Just make the breaking change, make it clear in the ChangeLog and let people upgrade.
2. Do a little dance with alternative names, and stick `{-# DEPRECATED #-}` warning on the old name to guide people through a migration path across multiple versions.
3. Create the function as `mRuntime` and then have `mRuntimeNoContext :: (Constraints) -> (event -> m result) -> m ()`. This is inconsistent with all the others (`ioRuntime` etc.)

I think (1) may be the best solution here, since the fix for existing code is so small.

I would consider deprecating `readerTRuntime` etc in favour of some haddock describing how to introduce monad transformers or effects. If you want use `ReaderT`, I think the clearest way is:
```haskell
mRuntimeWithContext $ \context event -> flip runReaderT context $ (yourCode :: ReaderT LambdaContext m result)
```
The library user can then also add `event` to the `ReaderT`, if desired.

Reply:
My thinking at this point is nearly "all of the above."  So:

  - Add `mRuntimeWithContext'`, deprecate `mRuntimeWithContext`, and warn that `mRuntimeWitContext` will replace it.
   - Deprecate `readerTRuntime`
   - Add `mRuntime` which has no `LambdaContext` anywhere
   - Possibly deprecate `HasLambdaContext`?
   - Eventually, break!

A weird bonus might be that this might make it easier to make `LambdaContext` construction lazy, so that no work is done if not needed, which also means it would partially address the current sam local incompatibilities (as it blows up due to incomplete context inputs--even if you _don't use_ the context).

My last hesitation here was that this lib just constantly makes Profunctor/Arrow pop up as an interesting abstraction, so I've always wanted to explore that (tho, probably in a separate lib).  Where does the context go in that case?  But it hit me, that that's easy too:  `SomeArrow (LambdaContext, event) output`.  And equally easy to remove the context if you're not interested.

I like to keep PRs as focused as possible, so my thinking is that this general approach is still right.  Step one, reduce envy's usage for greater compatibility, mark DefConfig deprecated (TODO), then remove it entirely when the `mRuntimeWithContext` break happens.

This all seems like a pretty nice balance of support while getting a positive trajectory.  What do you think?

Reply:
This all sounds pretty good to me. I think if you aren't using `HasLambdaContext` to stuff the context in, it's probably not worth keeping it around. Maybe deprecate it and see if it has use cases once you get through all the non-breaking parts of the refactoring? My only quibble is that `mRuntime`, if it kept with the naming conventions of all the other functions, would be `mRuntimeWithContext` but without the context. Maybe call it some obviously-temporary name like `mFullRuntime` or something, and flag that it's a temp name for the transition?

Re: Lazy context decoding. Anything that gets `sam local` working looks pretty good to me. I'd be interested to see how that laziness experiment turns out.

Re: Profunctor/Arrow: I imagine what you end up accepting is a runtime function of like `Kleisli m event result` (possibly with the context on the input too). I suspect you'll find wrangling arrows less comfortable than wrangling the `m` Monad, but it could be a good experiment for an add-on lib, perhaps.

Reply:
Oh, I think we’re on the same page with `mRuntime`.

ignoring constraints, I was thinking this version adds:
- `mRuntimeWithContext' :: (LambdaContext -> e -> m r) -> m ()`
- `mRuntime :: (e -> m r) -> m ()`

Where `mRuntime` remains permanent, and `mRuntimeWithContext'` eventually gets promoted to dropping the `'`, as the ReaderT based runtimes a dropped.

Already started working on these changes :)

Reply:
Yes, that sounds right to me.